File: application/backend-credit-card-enrollment/backend-java/src/main/java/cloud/ambar/common/commandhandler/CommandController.java
package cloud.ambar.common.commandhandler;

import cloud.ambar.common.projection.MongoTransactionalProjectionOperator;
import cloud.ambar.common.eventstore.PostgresTransactionalEventStore;
import lombok.RequiredArgsConstructor;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.util.Arrays;
import java.util.stream.Collectors;

@RequiredArgsConstructor
public class CommandController {
    private final PostgresTransactionalEventStore postgresTransactionalEventStore;
    private final MongoTransactionalProjectionOperator mongoTransactionalProjectionOperator;
    private static final Logger log = LogManager.getLogger(CommandController.class);

    public void processCommand(
            final Command command,
            final CommandHandler commandHandler
    ) {
        log.info("Command controller received command: " + command);

        try {
            // We start a PG transaction because command handlers need to append to the event store transactionally.
            // I.e., they need to read aggregates and append to them in an ACID fashion.
            // We start a Mongo transaction because if a command handler needs to read from a projection,
            // it also needs to do so transactionally to not receive dirty reads.
            postgresTransactionalEventStore.beginTransaction();
            mongoTransactionalProjectionOperator.startTransaction();
            commandHandler.handleCommand(command);
            postgresTransactionalEventStore.commitTransaction();
            mongoTransactionalProjectionOperator.commitTransaction();

            postgresTransactionalEventStore.abortDanglingTransactionsAndReturnConnectionToPool();
            mongoTransactionalProjectionOperator.abortDanglingTransactionsAndReturnSessionToPool();
        } catch (Exception e) {
            log.error("Failed to process reaction command.");
            log.error(e);
            log.error(e.getMessage());

            String stackTraceString = Arrays.stream(e.getStackTrace())
                    .map(StackTraceElement::toString)
                    .collect(Collectors.joining("\n"));
            log.error(stackTraceString);

            postgresTransactionalEventStore.abortDanglingTransactionsAndReturnConnectionToPool();
            mongoTransactionalProjectionOperator.abortDanglingTransactionsAndReturnSessionToPool();

            throw new RuntimeException("Failed to process command with exception: " + e);
        }
    }
}

File: application/backend-credit-card-enrollment/backend-c#/Common/Command/CommandController.cs
using CreditCardEnrollment.Common.EventStore;
using CreditCardEnrollment.Common.Projection;

namespace CreditCardEnrollment.Common.Command;

public class CommandController {
    private readonly PostgresTransactionalEventStore _postgresTransactionalEventStore;
    private readonly MongoTransactionalProjectionOperator _mongoTransactionalProjectionOperator;
    private readonly ILogger<CommandController> _logger;

    public CommandController(
        PostgresTransactionalEventStore postgresTransactionalEventStore,
        MongoTransactionalProjectionOperator mongoTransactionalProjectionOperator,
        ILogger<CommandController> logger
    ) {
        _postgresTransactionalEventStore = postgresTransactionalEventStore;
        _mongoTransactionalProjectionOperator = mongoTransactionalProjectionOperator;
        _logger = logger;
    }

    protected void ProcessCommand(Command command, CommandHandler commandHandler) {
        try {
            _logger.LogDebug("Starting to process command: {CommandType}", command.GetType().Name);
            _postgresTransactionalEventStore.BeginTransaction();
            _mongoTransactionalProjectionOperator.StartTransaction();
            commandHandler.HandleCommand(command);
            _postgresTransactionalEventStore.CommitTransaction();
            _mongoTransactionalProjectionOperator.CommitTransaction();

            _postgresTransactionalEventStore.AbortDanglingTransactionsAndReturnConnectionToPool();
            _mongoTransactionalProjectionOperator.AbortDanglingTransactionsAndReturnSessionToPool();
            _logger.LogDebug("Successfully processed command: {CommandType}", command.GetType().Name);
        } catch (Exception ex) {
            _postgresTransactionalEventStore.AbortDanglingTransactionsAndReturnConnectionToPool();
            _mongoTransactionalProjectionOperator.AbortDanglingTransactionsAndReturnSessionToPool();
            _logger.LogError("Exception in ProcessCommand: {0}, {1}", ex.Message, ex.StackTrace);
            throw new Exception($"Failed to process command: {ex.Message}", ex);
        }
    }
}